File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/EfficientOfflineAccumulator.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

/// Drop-in replacement for the mutable-var accumulation pattern that copies
/// the entire sample array on every sub-recording (`existingData + newSamples`).
/// This class appends to a single mutable array per data type, turning O(n²)
/// accumulation into O(n).
///
/// Self-contained: accepts raw `OfflineRecordingData<Any>` and does its own
/// PMD-to-Polar mapping internally, so callers need only a single method call.
internal class EfficientOfflineAccumulator {

    // MARK: - ACC
    private var accSamples: PolarAccData = []
    private var accStartTime: Date?
    private var accSettings: PolarSensorSetting?

    // MARK: - GYRO
    private var gyroSamples: PolarGyroData = []
    private var gyroStartTime: Date?
    private var gyroSettings: PolarSensorSetting?

    // MARK: - MAG
    private var magSamples: PolarMagnetometerData = []
    private var magStartTime: Date?
    private var magSettings: PolarSensorSetting?

    // MARK: - PPG
    private var ppgSamples: [(timeStamp: UInt64, channelSamples: [Int32], statusBits: [Int8]?)] = []
    private var ppgType: PpgDataType?
    private var ppgStartTime: Date?
    private var ppgSettings: PolarSensorSetting?

    // MARK: - PPI
    private var ppiSamples: [(timeStamp: UInt64, hr: Int, ppInMs: UInt16, ppErrorEstimate: UInt16, blockerBit: Int, skinContactStatus: Int, skinContactSupported: Int)] = []
    private var ppiStartTime: Date?

    // MARK: - HR
    private var hrSamples: PolarHrData = []
    private var hrStartTime: Date?

    // MARK: - Temperature
    private var tempSamples: [(timeStamp: UInt64, temperature: Float)] = []
    private var tempStartTime: Date?

    // MARK: - Skin Temperature
    private var skinTempSamples: [(timeStamp: UInt64, temperature: Float)] = []
    private var skinTempStartTime: Date?

    // MARK: - Empty
    private var emptyStartTime: Date?

    // MARK: - Accumulate

    /// Processes a single sub-recording chunk, mapping PMD types to Polar types
    /// and appending samples to the internal mutable arrays.
    func accumulate(_ offlineRecordingData: OfflineRecordingData<Any>) throws {
        let settings = offlineRecordingData.recordingSettings?.mapToPolarSettings() ?? PolarSensorSetting()
        let startTime = offlineRecordingData.startTime

        switch offlineRecordingData.data {
        case let accData as AccData:
            if accStartTime == nil {
                accStartTime = startTime
                accSettings = settings
            }
            for sample in accData.samples {
                accSamples.append((timeStamp: sample.timeStamp, x: sample.x, y: sample.y, z: sample.z))
            }

        case let gyroData as GyrData:
            if gyroStartTime == nil {
                gyroStartTime = startTime
                gyroSettings = settings
            }
            for sample in gyroData.samples {
                gyroSamples.append((timeStamp: sample.timeStamp, x: sample.x, y: sample.y, z: sample.z))
            }

        case let magData as MagData:
            if magStartTime == nil {
                magStartTime = startTime
                magSettings = settings
            }
            for sample in magData.samples {
                magSamples.append((timeStamp: sample.timeStamp, x: sample.x, y: sample.y, z: sample.z))
            }

        case let ppgData as PpgData:
            if ppgStartTime == nil {
                ppgStartTime = startTime
                ppgSettings = settings
            }
            for sample in ppgData.samples {
                if sample.frameType == PmdDataFrameType.type_0 {
                    let d = sample as! PpgDataFrameType0
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: [d.ppgDataSamples[0], d.ppgDataSamples[1], d.ppgDataSamples[2], d.ambientSample], statusBits: nil))
                    ppgType = .ppg3_ambient1
                } else if sample.frameType == PmdDataFrameType.type_6 {
                    let d = sample as! PpgDataFrameType6
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: [d.sportId], statusBits: nil))
                    ppgType = .ppg1
                } else if sample.frameType == PmdDataFrameType.type_7 {
                    let d = sample as! PpgDataFrameType7
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: d.ppgDataSamples, statusBits: nil))
                    ppgType = .ppg17
                } else if sample.frameType == PmdDataFrameType.type_10 {
                    let d = sample as! PpgDataFrameType10
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: d.greenSamples + d.redSamples + d.irSamples, statusBits: d.statusBits))
                    ppgType = .ppg21
                } else if sample.frameType == PmdDataFrameType.type_9 {
                    let d = sample as! PpgDataFrameType9
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: d.ppgDataSamples, statusBits: nil))
                    ppgType = .ppg3
                } else if sample.frameType == PmdDataFrameType.type_13 {
                    let d = sample as! PpgDataFrameType13
                    ppgSamples.append((timeStamp: sample.timeStamp!, channelSamples: [d.ppgDataSamples[0], d.ppgDataSamples[1]], statusBits: d.statusBits))
                    ppgType = .ppg2
                }
            }

        case let ppiData as PpiData:
            if ppiStartTime == nil {
                ppiStartTime = startTime
            }
            for sample in ppiData.samples {
                ppiSamples.append((timeStamp: sample.timeStamp, hr: sample.hr, ppInMs: sample.ppInMs, ppErrorEstimate: sample.ppErrorEstimate, blockerBit: sample.blockerBit, skinContactStatus: sample.skinContactStatus, skinContactSupported: sample.skinContactSupported))
            }

        case let hrData as OfflineHrData:
            if hrStartTime == nil {
                hrStartTime = startTime
            }
            for sample in hrData.samples {
                hrSamples.append((hr: sample.hr, ppgQuality: sample.ppgQuality, correctedHr: sample.correctedHr, rrsMs: [], rrAvailable: false, contactStatus: false, contactStatusSupported: false))
            }

        case let temperatureData as TemperatureData:
            if tempStartTime == nil {
                tempStartTime = startTime
            }
            for sample in temperatureData.samples {
                tempSamples.append((timeStamp: sample.timeStamp, temperature: sample.temperature))
            }

        case let skinTemperatureData as SkinTemperatureData:
            if skinTempStartTime == nil {
                skinTempStartTime = startTime
            }
            for sample in skinTemperatureData.samples {
                skinTempSamples.append((timeStamp: sample.timeStamp, temperature: sample.skinTemperature))
            }

        case _ as EmptyData:
            if emptyStartTime == nil {
                emptyStartTime = startTime
            }

        default:
            throw PolarErrors.polarOfflineRecordingError(
                description: "GetOfflineRecording failed. Data type is not supported."
            )
        }
    }

    // MARK: - Result

    func getResult() -> PolarOfflineRecordingData? {
        if let startTime = ppiStartTime, !ppiSamples.isEmpty {
            return .ppiOfflineRecordingData(
                (timeStamp: UInt64(startTime.timeIntervalSince1970), samples: ppiSamples),
                startTime: startTime
            )
        }
        if let startTime = ppgStartTime, let settings = ppgSettings, let type = ppgType, !ppgSamples.isEmpty {
            return .ppgOfflineRecordingData(
                (type: type, samples: ppgSamples),
                startTime: startTime,
                settings: settings
            )
        }
        if let startTime = accStartTime, let settings = accSettings, !accSamples.isEmpty {
            return .accOfflineRecordingData(accSamples, startTime: startTime, settings: settings)
        }
        if let startTime = gyroStartTime, let settings = gyroSettings, !gyroSamples.isEmpty {
            return .gyroOfflineRecordingData(gyroSamples, startTime: startTime, settings: settings)
        }
        if let startTime = magStartTime, let settings = magSettings, !magSamples.isEmpty {
            return .magOfflineRecordingData(magSamples, startTime: startTime, settings: settings)
        }
        if let startTime = hrStartTime, !hrSamples.isEmpty {
            return .hrOfflineRecordingData(hrSamples, startTime: startTime)
        }
        if let startTime = tempStartTime, !tempSamples.isEmpty {
            return .temperatureOfflineRecordingData(
                (timeStamp: tempSamples.last?.timeStamp ?? 0, samples: tempSamples),
                startTime: startTime
            )
        }
        if let startTime = skinTempStartTime, !skinTempSamples.isEmpty {
            return .skinTemperatureOfflineRecordingData(
                (timeStamp: skinTempSamples.last?.timeStamp ?? 0, samples: skinTempSamples),
                startTime: startTime
            )
        }
        if let startTime = emptyStartTime {
            return .emptyData(startTime: startTime)
        }
        return nil
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/EfficientOfflineAccumulator.kt
// Copyright © 2024 Polar Electro Oy. All rights reserved.
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model.*
import com.polar.androidcommunications.api.ble.model.offlinerecording.OfflineRecordingData
import com.polar.sdk.api.errors.PolarOfflineRecordingError
import com.polar.sdk.api.model.*
import com.polar.sdk.impl.utils.PolarDataUtils.mapPMDClientOfflineHrDataToPolarHrData
import com.polar.sdk.impl.utils.PolarDataUtils.mapPMDClientOfflineTemperatureDataToPolarTemperatureData
import com.polar.sdk.impl.utils.PolarDataUtils.mapPMDClientPpgDataToPolarPpg
import com.polar.sdk.impl.utils.PolarDataUtils.mapPMDClientPpiDataToPolarPpiData
import com.polar.sdk.impl.utils.PolarDataUtils.mapPmdClientAccDataToPolarAcc
import com.polar.sdk.impl.utils.PolarDataUtils.mapPmdClientGyroDataToPolarGyro
import com.polar.sdk.impl.utils.PolarDataUtils.mapPmdClientMagDataToPolarMagnetometer
import com.polar.sdk.impl.utils.PolarDataUtils.mapPmdClientSkinTemperatureDataToPolarTemperatureData
import com.polar.sdk.impl.utils.PolarDataUtils.mapPmdSettingsToPolarSettings
import java.time.LocalDateTime

/**
 * Drop-in replacement for OfflineRecordingAccumulator that appends samples
 * to a single mutable list instead of re-copying the entire list on every
 * sub-recording. Turns O(n²) accumulation into O(n).
 */
internal class EfficientOfflineAccumulator {

    // ACC
    private var accSamples: MutableList<PolarAccelerometerData.PolarAccelerometerDataSample>? = null
    private var accSettings: PolarSensorSetting? = null
    private var accStartTime: LocalDateTime? = null

    // GYRO
    private var gyroSamples: MutableList<PolarGyroData.PolarGyroDataSample>? = null
    private var gyroSettings: PolarSensorSetting? = null
    private var gyroStartTime: LocalDateTime? = null

    // MAG
    private var magSamples: MutableList<PolarMagnetometerData.PolarMagnetometerDataSample>? = null
    private var magSettings: PolarSensorSetting? = null
    private var magStartTime: LocalDateTime? = null

    // PPG
    private var ppgSamples: MutableList<PolarPpgData.PolarPpgSample>? = null
    private var ppgType: PolarPpgData.PpgDataType? = null
    private var ppgSettings: PolarSensorSetting? = null
    private var ppgStartTime: LocalDateTime? = null

    // PPI
    private var ppiSamples: MutableList<PolarPpiData.PolarPpiSample>? = null
    private var ppiStartTime: LocalDateTime? = null

    // HR
    private var hrSamples: MutableList<PolarHrData.PolarHrSample>? = null
    private var hrStartTime: LocalDateTime? = null

    // Temperature
    private var tempSamples: MutableList<PolarTemperatureData.PolarTemperatureDataSample>? = null
    private var tempStartTime: LocalDateTime? = null

    // Skin temperature
    private var skinTempSamples: MutableList<PolarTemperatureData.PolarTemperatureDataSample>? = null
    private var skinTempStartTime: LocalDateTime? = null

    fun accumulate(offlineRecData: OfflineRecordingData<*>) {
        when (val data = offlineRecData.data) {
            is AccData -> {
                val polar = mapPmdClientAccDataToPolarAcc(data)
                if (accSamples == null) {
                    accSamples = polar.samples.toMutableList()
                    accSettings = offlineRecData.recordingSettings?.let {
                        mapPmdSettingsToPolarSettings(it, fromSelected = false)
                    } ?: throw PolarOfflineRecordingError("getOfflineRecord failed. Acc data is missing settings")
                    accStartTime = offlineRecData.startTime
                } else {
                    accSamples!!.addAll(polar.samples)
                }
            }

            is GyrData -> {
                val polar = mapPmdClientGyroDataToPolarGyro(data)
                if (gyroSamples == null) {
                    gyroSamples = polar.samples.toMutableList()
                    gyroSettings = offlineRecData.recordingSettings?.let {
                        mapPmdSettingsToPolarSettings(it, fromSelected = false)
                    } ?: throw PolarOfflineRecordingError("getOfflineRecord failed. Gyro data is missing settings")
                    gyroStartTime = offlineRecData.startTime
                } else {
                    gyroSamples!!.addAll(polar.samples)
                }
            }

            is MagData -> {
                val polar = mapPmdClientMagDataToPolarMagnetometer(data)
                if (magSamples == null) {
                    magSamples = polar.samples.toMutableList()
                    magSettings = offlineRecData.recordingSettings?.let {
                        mapPmdSettingsToPolarSettings(it, fromSelected = false)
                    } ?: throw PolarOfflineRecordingError("getOfflineRecord failed. Magnetometer data is missing settings")
                    magStartTime = offlineRecData.startTime
                } else {
                    magSamples!!.addAll(polar.samples)
                }
            }

            is PpgData -> {
                val polar = mapPMDClientPpgDataToPolarPpg(data)
                if (ppgSamples == null) {
                    ppgSamples = polar.samples.toMutableList()
                    ppgType = polar.type
                    ppgSettings = offlineRecData.recordingSettings?.let {
                        mapPmdSettingsToPolarSettings(it, fromSelected = false)
                    } ?: throw PolarOfflineRecordingError("getOfflineRecord failed. Ppg data is missing settings")
                    ppgStartTime = offlineRecData.startTime
                } else {
                    ppgSamples!!.addAll(polar.samples)
                    ppgType = polar.type
                }
            }

            is PpiData -> {
                val polar = mapPMDClientPpiDataToPolarPpiData(data)
                if (ppiSamples == null) {
                    ppiSamples = polar.samples.toMutableList()
                    ppiStartTime = offlineRecData.startTime
                } else {
                    ppiSamples!!.addAll(polar.samples)
                }
            }

            is OfflineHrData -> {
                val polar = mapPMDClientOfflineHrDataToPolarHrData(data)
                if (hrSamples == null) {
                    hrSamples = polar.samples.toMutableList()
                    hrStartTime = offlineRecData.startTime
                } else {
                    hrSamples!!.addAll(polar.samples)
                }
            }

            is TemperatureData -> {
                val polar = mapPMDClientOfflineTemperatureDataToPolarTemperatureData(data)
                if (tempSamples == null) {
                    tempSamples = polar.samples.toMutableList()
                    tempStartTime = offlineRecData.startTime
                } else {
                    tempSamples!!.addAll(polar.samples)
                }
            }

            is SkinTemperatureData -> {
                val polar = mapPmdClientSkinTemperatureDataToPolarTemperatureData(data)
                if (skinTempSamples == null) {
                    skinTempSamples = polar.samples.toMutableList()
                    skinTempStartTime = offlineRecData.startTime
                } else {
                    skinTempSamples!!.addAll(polar.samples)
                }
            }

            else -> throw PolarOfflineRecordingError("Data type is not supported.")
        }
    }

    fun getResult(): PolarOfflineRecordingData? {
        ppiSamples?.let {
            return PolarOfflineRecordingData.PpiOfflineRecording(PolarPpiData(it), ppiStartTime!!)
        }
        ppgSamples?.let {
            return PolarOfflineRecordingData.PpgOfflineRecording(PolarPpgData(it, ppgType!!), ppgStartTime!!, ppgSettings)
        }
        accSamples?.let {
            return PolarOfflineRecordingData.AccOfflineRecording(PolarAccelerometerData(it), accStartTime!!, accSettings!!)
        }
        gyroSamples?.let {
            return PolarOfflineRecordingData.GyroOfflineRecording(PolarGyroData(it), gyroStartTime!!, gyroSettings!!)
        }
        magSamples?.let {
            return PolarOfflineRecordingData.MagOfflineRecording(PolarMagnetometerData(it), magStartTime!!, magSettings)
        }
        hrSamples?.let {
            return PolarOfflineRecordingData.HrOfflineRecording(PolarHrData(it), hrStartTime!!)
        }
        tempSamples?.let {
            return PolarOfflineRecordingData.TemperatureOfflineRecording(PolarTemperatureData(it), tempStartTime!!)
        }
        skinTempSamples?.let {
            return PolarOfflineRecordingData.SkinTemperatureOfflineRecording(PolarTemperatureData(it), skinTempStartTime!!)
        }
        return null
    }
}
